File: website/src/games/dartArenaScene.ts
import type {PixelColor} from './pixelTypes';
import {PixelContext, hash} from './pixelContext';
import {GRID_HEIGHT, GRID_WIDTH} from './pixelTypes';

const w = GRID_WIDTH;
const h = GRID_HEIGHT;

type Side = 'left' | 'right';

/** Port `Games/DartArenaScene.swift` */
export function renderDartArena(
  ctx: PixelContext,
  boardCenterX: number,
  boardCenterY: number,
  boardRadius: number,
  animTick: number,
): void {
  fillBackground(ctx);
  drawBackWall(ctx, boardCenterX, boardCenterY, boardRadius);
  drawCeiling(ctx, boardCenterX, animTick);
  drawSpotlights(ctx, boardCenterX, boardCenterY);
  drawSideStands(ctx, 'left', animTick);
  drawSideStands(ctx, 'right', animTick);
  drawFloor(ctx, boardCenterX, boardCenterY, boardRadius);
  drawBoardGlow(ctx, boardCenterX, boardCenterY, boardRadius);
}

function fillBackground(ctx: PixelContext): void {
  ctx.rect(0, 0, w, h, 'black');
  for (let y = 0; y < h; y++) {
    const t = y / (h - 1);
    const band: PixelColor = y < 12 ? 'black' : t < 0.55 ? 'navy' : 'road';
    ctx.rect(0, y, w, 1, band);
  }
}

function drawBackWall(
  ctx: PixelContext,
  boardCenterX: number,
  boardCenterY: number,
  boardRadius: number,
): void {
  const wallBottom = Math.floor(boardCenterY - boardRadius - 4);
  if (wallBottom <= 8) return;
  for (let y = 8; y < wallBottom; y++) {
    for (let x = 0; x < w; x++) {
      const dx = Math.abs(x - boardCenterX) / boardCenterX;
      const shade: PixelColor = hash(x, y) % 5 === 0 ? 'navy' : 'darkGray';
      const depth = 1 - (y / wallBottom) * 0.35;
      if (dx < 0.92 * depth) {
        ctx.rect(x, y, 1, 1, shade);
      }
    }
  }
  ctx.rect(0, wallBottom, w, 2, 'wood');
}

function drawCeiling(ctx: PixelContext, boardCenterX: number, animTick: number): void {
  for (let x = 0; x < w; x += 18) {
    ctx.rect(x, 2, 1, 6, 'darkGray');
    ctx.rect(x, 8, 14, 1, 'darkGray');
  }
  const flicker = Math.floor(animTick / 18) % 2 === 0;
  for (const lx of [24, 56, 80, 104, 132]) {
    const glow: PixelColor =
      flicker && hash(lx, animTick) % 3 === 0 ? 'yellow' : 'white';
    ctx.rect(lx, 4, 3, 2, glow);
    ctx.rect(lx + 1, 6, 1, 1, 'yellow');
  }
  ctx.rect(Math.round(boardCenterX) - 2, 3, 5, 2, 'cyan');
}

function drawSpotlights(
  ctx: PixelContext,
  boardCenterX: number,
  boardCenterY: number,
): void {
  const targetY = Math.round(boardCenterY);
  const targetX = Math.round(boardCenterX);
  for (const originX of [12, 148]) {
    drawLightBeam(ctx, originX, 6, targetX, targetY, 'cyan');
  }
  drawLightBeam(ctx, Math.round(boardCenterX), 4, targetX, targetY, 'yellow');
}

function drawLightBeam(
  ctx: PixelContext,
  fromX: number,
  fromY: number,
  toX: number,
  toY: number,
  color: PixelColor,
): void {
  const steps = Math.max(Math.abs(toX - fromX), Math.abs(toY - fromY));
  if (steps <= 0) return;
  for (let i = 0; i < steps; i += 2) {
    const t = i / steps;
    const x = fromX + Math.floor((toX - fromX) * t);
    const y = fromY + Math.floor((toY - fromY) * t);
    if (hash(x, y) % 4 !== 0) continue;
    ctx.rect(x, y, 1, 1, color);
  }
}

function drawSideStands(ctx: PixelContext, side: Side, animTick: number): void {
  const rows = 10;
  for (let row = 0; row < rows; row++) {
    const depth = row;
    const y = 12 + row * 6;
    if (y >= h - 14) continue;
    const tierWidth = 38 - depth * 2;
    const x0 = side === 'left' ? 1 + depth : w - tierWidth - 1 - depth;
    ctx.rect(x0, y - 1, tierWidth, 1, 'wood');
    ctx.rect(x0, y + 4, tierWidth, 1, 'white');
    for (let col = 0; col < tierWidth; col++) {
      const x = x0 + col;
      if (shouldDrawCrowdPixel(x, y, row, col, animTick)) {
        ctx.rect(x, y, 1, 1, crowdColor(x, y, animTick));
        if (row > 2 && hash(x, y + 1) % 3 === 0) {
          ctx.rect(x, y + 1, 1, 1, 'darkGray');
        }
      }
    }
  }
  const pillarX = side === 'left' ? 42 : 115;
  for (let py = 10; py < h - 12; py++) {
    if (py % 7 === 0) continue;
    ctx.rect(pillarX, py, 2, 1, 'wood');
  }
}

function shouldDrawCrowdPixel(
  x: number,
  y: number,
  row: number,
  col: number,
  animTick: number,
): boolean {
  const wave = Math.floor(animTick / 6 + row);
  const base = hash(x + wave, y) % 10;
  if (base < 2) return false;
  const cheer = (animTick + col) % 24 < 4;
  return cheer ? base < 9 : base < 7;
}

function crowdColor(x: number, y: number, animTick: number): PixelColor {
  const palette: PixelColor[] = ['cyan', 'magenta', 'yellow', 'green', 'red', 'white'];
  return palette[hash(x + Math.floor(animTick / 10), y) % palette.length];
}

function drawFloor(
  ctx: PixelContext,
  boardCenterX: number,
  boardCenterY: number,
  boardRadius: number,
): void {
  const floorTop = Math.round(boardCenterY + boardRadius + 2);
  if (floorTop >= h - 2) return;
  const vanishX = boardCenterX;

  for (let y = floorTop; y < h; y++) {
    const t = (y - floorTop) / (h - floorTop);
    const halfWidth = 12 + t * 78;
    const left = Math.floor(vanishX - halfWidth);
    const right = Math.floor(vanishX + halfWidth);
    for (let x = Math.max(0, left); x <= Math.min(w - 1, right); x++) {
      const u = (x - left) / Math.max(1, right - left);
      const plank = Math.floor(u * 14) % 2 === 0;
      const shade: PixelColor = plank ? 'wood' : 'road';
      if (hash(x, y) % 11 === 0) {
        ctx.rect(x, y, 1, 1, 'darkGray');
      } else {
        ctx.rect(x, y, 1, 1, shade);
      }
    }
    if (y % 5 === 0) {
      const lineLeft = Math.floor(vanishX - halfWidth * 0.92);
      const lineRight = Math.floor(vanishX + halfWidth * 0.92);
      ctx.rect(Math.max(0, lineLeft), y, Math.max(1, lineRight - lineLeft), 1, 'black');
    }
  }
  ctx.rect(0, floorTop, w, 1, 'yellow');
}

function drawBoardGlow(
  ctx: PixelContext,
  boardCenterX: number,
  boardCenterY: number,
  boardRadius: number,
): void {
  const cx = Math.round(boardCenterX);
  const cy = Math.round(boardCenterY);
  const r = Math.ceil(boardRadius) + 3;
  for (let py = cy - r; py <= cy + r; py++) {
    for (let px = cx - r; px <= cx + r; px++) {
      const dx = px - cx;
      const dy = py - cy;
      const d = Math.sqrt(dx * dx + dy * dy);
      if (d > boardRadius && d < boardRadius + 3.5 && hash(px, py) % 2 === 0) {
        ctx.rect(px, py, 1, 1, 'yellow');
      }
    }
  }
}

File: app/Games/DartArenaScene.swift
import Foundation

/// Pixelowa hala dartowa (widok 3D) — publiczność z boków, światła, parkiet.
enum DartArenaScene {
  private static let w = GameContext.width
  private static let h = GameContext.height

  /// Renderuje tło i otoczenie areny wokół tarczy dartowej.
  ///
  /// - Parameters:
  ///   - context: Docelowy kontekst rasteryzacji pikselowej.
  ///   - boardCenterX: Pozycja X środka tarczy.
  ///   - boardCenterY: Pozycja Y środka tarczy.
  ///   - boardRadius: Promień tarczy.
  ///   - animTick: Licznik animacji używany do subtelnych efektów sceny.
  static func render(
    context: GameContext,
    boardCenterX: Double,
    boardCenterY: Double,
    boardRadius: Double,
    animTick: UInt
  ) {
    fillBackground(context: context)
    drawBackWall(context: context, boardCenterX: boardCenterX, boardCenterY: boardCenterY, boardRadius: boardRadius)
    drawCeiling(context: context, boardCenterX: boardCenterX, animTick: animTick)
    drawSpotlights(context: context, boardCenterX: boardCenterX, boardCenterY: boardCenterY)
    drawSideStands(context: context, side: .left, animTick: animTick)
    drawSideStands(context: context, side: .right, animTick: animTick)
    drawFloor(context: context, boardCenterX: boardCenterX, boardCenterY: boardCenterY, boardRadius: boardRadius)
    drawBoardGlow(context: context, boardCenterX: boardCenterX, boardCenterY: boardCenterY, boardRadius: boardRadius)
  }

  private enum Side { case left, right }

  private static func fillBackground(context: GameContext) {
    context.rect(x: 0, y: 0, width: w, height: h, color: .black)
    for y in 0..<h {
      let t = Double(y) / Double(h - 1)
      let band = y < 12 ? PixelColor.black : (t < 0.55 ? PixelColor.navy : PixelColor.road)
      context.rect(x: 0, y: y, width: w, height: 1, color: band)
    }
  }

  private static func drawBackWall(
    context: GameContext,
    boardCenterX: Double,
    boardCenterY: Double,
    boardRadius: Double
  ) {
    let wallBottom = Int((boardCenterY - boardRadius - 4).rounded(.down))
    guard wallBottom > 8 else { return }
    for y in 8..<wallBottom {
      for x in 0..<w {
        let dx = abs(Double(x) - boardCenterX) / boardCenterX
        let shade: PixelColor = hash(x, y) % 5 == 0 ? .navy : .darkGray
        let depth = 1.0 - Double(y) / Double(wallBottom) * 0.35
        if dx < 0.92 * depth {
          context.rect(x: x, y: y, width: 1, height: 1, color: shade)
        }
      }
    }
    // Pas startowy za tarczą
    let bandY = wallBottom
    context.rect(x: 0, y: bandY, width: w, height: 2, color: .wood)
  }

  private static func drawCeiling(context: GameContext, boardCenterX: Double, animTick: UInt) {
    for x in stride(from: 0, to: w, by: 18) {
      context.rect(x: x, y: 2, width: 1, height: 6, color: .darkGray)
      context.rect(x: x, y: 8, width: 14, height: 1, color: .darkGray)
    }
    let flicker = (animTick / 18) % 2 == 0
    for lx in [24, 56, 80, 104, 132] {
      let glow = flicker && hash(lx, Int(animTick)) % 3 == 0 ? PixelColor.yellow : PixelColor.white
      context.rect(x: lx, y: 4, width: 3, height: 2, color: glow)
      context.rect(x: lx + 1, y: 6, width: 1, height: 1, color: .yellow)
    }
    // Reflektor centralny
    context.rect(x: Int(boardCenterX) - 2, y: 3, width: 5, height: 2, color: .cyan)
  }

  private static func drawSpotlights(
    context: GameContext,
    boardCenterX: Double,
    boardCenterY: Double
  ) {
    let targetY = Int(boardCenterY.rounded())
    let targetX = Int(boardCenterX.rounded())
    for originX in [12, 148] {
      drawLightBeam(
        context: context,
        fromX: originX,
        fromY: 6,
        toX: targetX,
        toY: targetY,
        color: .cyan
      )
    }
    drawLightBeam(
      context: context,
      fromX: Int(boardCenterX),
      fromY: 4,
      toX: targetX,
      toY: targetY,
      color: .yellow
    )
  }

  private static func drawLightBeam(
    context: GameContext,
    fromX: Int,
    fromY: Int,
    toX: Int,
    toY: Int,
    color: PixelColor
  ) {
    let steps = max(abs(toX - fromX), abs(toY - fromY))
    guard steps > 0 else { return }
    for i in stride(from: 0, to: steps, by: 2) {
      let t = Double(i) / Double(steps)
      let x = fromX + Int(Double(toX - fromX) * t)
      let y = fromY + Int(Double(toY - fromY) * t)
      if hash(x, y) % 4 != 0 { continue }
      context.rect(x: x, y: y, width: 1, height: 1, color: color)
    }
  }

  private static func drawSideStands(context: GameContext, side: Side, animTick: UInt) {
    let rows = 10
    for row in 0..<rows {
      let depth = row
      let y = 12 + row * 6
      guard y < h - 14 else { continue }
      let tierWidth = 38 - depth * 2
      let x0: Int
      switch side {
      case .left:
        x0 = 1 + depth
      case .right:
        x0 = w - tierWidth - 1 - depth
      }
      // Balustrada
      context.rect(x: x0, y: y - 1, width: tierWidth, height: 1, color: .wood)
      context.rect(x: x0, y: y + 4, width: tierWidth, height: 1, color: .white)

      for col in 0..<tierWidth {
        let x = x0 + col
        if shouldDrawCrowdPixel(x: x, y: y, row: row, col: col, animTick: animTick) {
          context.rect(x: x, y: y, width: 1, height: 1, color: crowdColor(x: x, y: y, animTick: animTick))
          if row > 2, hash(x, y + 1) % 3 == 0 {
            context.rect(x: x, y: y + 1, width: 1, height: 1, color: .darkGray)
          }
        }
      }
    }
    // Słupy areny
    let pillarX = side == .left ? 42 : 115
    for py in 10..<(h - 12) {
      if py % 7 == 0 { continue }
      context.rect(x: pillarX, y: py, width: 2, height: 1, color: .wood)
    }
  }

  private static func shouldDrawCrowdPixel(x: Int, y: Int, row: Int, col: Int, animTick: UInt) -> Bool {
    let wave = Int(animTick / 6 + UInt(row))
    let base = hash(x + wave, y) % 10
    if base < 2 { return false }
    let cheer = (animTick + UInt(col)) % 24 < 4
    return cheer ? base < 9 : base < 7
  }

  private static func crowdColor(x: Int, y: Int, animTick: UInt) -> PixelColor {
    let palette: [PixelColor] = [.cyan, .magenta, .yellow, .green, .red, .white]
    let i = hash(x + Int(animTick / 10), y) % palette.count
    return palette[i]
  }

  private static func drawFloor(
    context: GameContext,
    boardCenterX: Double,
    boardCenterY: Double,
    boardRadius: Double
  ) {
    let floorTop = Int((boardCenterY + boardRadius + 2).rounded())
  guard floorTop < h - 2 else { return }
    let vanishX = boardCenterX
    let vanishY = Double(floorTop) + 4

    for y in floorTop..<h {
      let t = Double(y - floorTop) / Double(h - floorTop)
      let halfWidth = 12 + t * 78
      let left = Int(vanishX - halfWidth)
      let right = Int(vanishX + halfWidth)
      for x in max(0, left)...min(w - 1, right) {
        let u = Double(x - left) / max(1, Double(right - left))
        let plank = Int(u * 14) % 2 == 0
        let shade: PixelColor = plank ? .wood : .road
        if hash(x, y) % 11 == 0 {
          context.rect(x: x, y: y, width: 1, height: 1, color: .darkGray)
        } else {
          context.rect(x: x, y: y, width: 1, height: 1, color: shade)
        }
      }
      if y % 5 == 0 {
        let lineLeft = Int(vanishX - halfWidth * 0.92)
        let lineRight = Int(vanishX + halfWidth * 0.92)
        context.rect(x: max(0, lineLeft), y: y, width: max(1, lineRight - lineLeft), height: 1, color: .black)
      }
    }

    // Ochełnienie ochepek (rzuca cień)
    let lineY = floorTop
    context.rect(x: 0, y: lineY, width: w, height: 1, color: .yellow)
  }

  private static func drawBoardGlow(
    context: GameContext,
    boardCenterX: Double,
    boardCenterY: Double,
    boardRadius: Double
  ) {
    let cx = Int(boardCenterX.rounded())
    let cy = Int(boardCenterY.rounded())
    let r = Int(boardRadius.rounded(.up)) + 3
    for py in (cy - r)...(cy + r) {
      for px in (cx - r)...(cx + r) {
        let dx = Double(px - cx)
        let dy = Double(py - cy)
        let d = sqrt(dx * dx + dy * dy)
        guard d > boardRadius, d < boardRadius + 3.5 else { continue }
        if hash(px, py) % 2 == 0 {
          context.rect(x: px, y: py, width: 1, height: 1, color: .yellow)
        }
      }
    }
  }

  private static func hash(_ a: Int, _ b: Int) -> Int {
    var x = UInt(bitPattern: a &* 374_761_393 &+ b &* 668_265_263)
    x = (x ^ (x >> 13)) &* 1_274_126_171
    return Int(x % 9_999)
  }
}
